test/common/http/hcm_router_fuzz_test.cc

```diff
@@ -0,0 +1,575 @@
+    }
+  }
+
+  void sendData(const std::string& request_data, bool end_stream) {
+    if (request_state_ == StreamState::PendingDataOrTrailers) {
+      request_state_ = end_stream ? StreamState::Closed : StreamState::PendingDataOrTrailers;
+      Buffer::OwnedImpl data(request_data);
+      decoder_->decodeData(data, end_stream);
+    }
+  }
+
+  void sendTrailers(const test::fuzz::Headers& request_trailers) {
+    if (request_state_ == StreamState::PendingDataOrTrailers) {
+      auto trailers = std::make_unique<TestRequestTrailerMapImpl>(
+          Fuzz::fromHeaders<TestRequestTrailerMapImpl>(request_trailers));
+      decoder_->decodeTrailers(std::move(trailers));
+      request_state_ = StreamState::Closed;
+    }
+  }
+
+  ConnectionManagerImpl& conn_manager_;
+  FuzzConfig& config_;
+  RequestDecoder* decoder_{};
+  NiceMock<MockResponseEncoder> encoder_;
+  StreamState request_state_{StreamState::PendingHeaders};
+};
+using FuzzDownstreamPtr = std::unique_ptr<FuzzDownstream>;
+
+// This class mocks the upstream and serves as entry point for responses.
+class FuzzUpstream {
+public:
+  FuzzUpstream(Http::ResponseDecoder& decoder) : decoder_(decoder) {
+    ON_CALL(mock_request_encoder_, encodeHeaders(_, _))
+        .WillByDefault(Invoke([](const Http::RequestHeaderMap&, bool) { return okStatus(); }));
+    ON_CALL(mock_request_encoder_.stream_, resetStream(_))
+        .WillByDefault(
+            Invoke([this](StreamResetReason) { response_state_ = StreamState::Closed; }));
+  }
+
+  void sendHeaders(Http::ResponseHeaderMapPtr&& headers, bool end_stream) {
+    if (response_state_ == StreamState::PendingHeaders) {
+      response_state_ = end_stream ? StreamState::Closed : StreamState::PendingDataOrTrailers;
+      decoder_.decodeHeaders(std::move(headers), end_stream);
+    }
+  }
+
+  void sendData(Buffer::Instance& data, bool end_stream) {
+    if (response_state_ == StreamState::PendingDataOrTrailers) {
+      response_state_ = end_stream ? StreamState::Closed : StreamState::PendingDataOrTrailers;
+      decoder_.decodeData(data, end_stream);
+    }
+  }
+
+  void sendTrailers(Http::ResponseTrailerMapPtr&& trailers) {
+    if (response_state_ == StreamState::PendingDataOrTrailers) {
+      response_state_ = StreamState::Closed;
+      decoder_.decodeTrailers(std::move(trailers));
+    }
+  }
+
+  Http::ResponseDecoder& decoder_;
+  StreamState response_state_{StreamState::PendingHeaders};
+  NiceMock<MockRequestEncoder> mock_request_encoder_;
+};
+
+// This class mocks an upstream cluster. It holds the set of
+// `FuzzUpstream` instances, which result from `FuzzDownstream`
+// requests
+class FuzzCluster {
+public:
+  FuzzCluster(FuzzConfig& cfg, const char* path, bool internal_redirect_policy_enabled,
+              bool cross_scheme_redirect_allowed, bool allows_early_data_for_request)
+      : cfg_(cfg), name_(path), path_(path), mock_route_(new Router::MockRoute()),
```

Comment:
This is a test code, so I guess it is less important, but `mock_route_` is allocated and never freed. Can this use a unique_ptr instead?

Reply:
`mock_route_` will be used to initialize the `shared_ptr route_`, which will eventually free `mock_route_`. I agree this is awkward, but the alternative is to cast `route_` back to a `MockRoute`. I have now opted to initialize these two in the constructor body in order to make it more readable.